feat: fix merge branch support on GitHub

doudou opened the pull request:
After the refactoring for GitLab support, GitHub's support to use merge branches was broken. The issue was that the GH API does not return the "mergeable" flag in its PR list endpoint, only on the per-pull request GET endpoint.

I started fixing this by looking for the merge_commit_sha field that does exist in the listing endpoint. Turns out that this is not enough, as GH does not re-compute the merge commit in some cases. We have to actually hit the pull request GET endpoint to force the update.

At that point, ended up having issues with faraday-cache. The GET endpoint was being cached, which meant that we were not triggering the merge commit computation and were waiting forever for it to be updated. I disabled caching for this endpoint - using a separate octokit client for it. Originally, I disabled for the whole daemon but that would be triggering rate limits regularly.

doudou committed:
fix: detection of mergeable/non-mergeable PR in the github service

GitHub's "Listing pull requests" endpoint does not return the mergeable
attribute, which leads right now to having all pull requests use the head
ref instead of the merge ref.

This fix is limited to github as I don't know whether gitlab has the
same issue.

doudou committed:
feat: make the strategy to choose the test branch configurable

I personally would like to force using the merge commit. Before I bit
the bullet and looked for it, the bug related to mergeability on github
made me waste a LOT of time checking why pull requests won't build,
to finally realize it's because things that were on master were actually
not included.

At least, if we always use 'merge' (except for drafts), I'm going to
get a clear update error that the merge commit does not exist.

doudou committed:
fix: set rubocop target version to 2.5

g-arjones commented:
Could you please explain what you are trying to improve in the end-user experience and/or your use case? I think it's reasonable to choose to always build the head commit but, in the context of the daemon, always building the merge commit sounds more like a bug to me since that might not be available or may not be what the user expects. I also don't understand why you override GitHub's mergeable status... I think you may build an outdated commit if a new push has conflicts with the base branch (in which case `mergeable` will be `false` and `merge_commit_sha` will point to the previous merge commit sha, if any).

doudou commented:
You can find explanation about all of this in the two commit messages. If you prefer, I can copy/paste both in the pull request message. But since there are only two, I thought I could leave it this way.

g-arjones commented:
> You can find explanation about all of this in the two commit messages

I missed that, sorry. I get it now.

doudou committed:
fix: detection of merge commit on GitHub

In some conditions, GitHub will not update the merge commit unless
explicitly queried, either by opening the pull request page on the
Web UI, or doing a GET on the specific pull request endpoint. In
the latter case, the mergeable attribute indicates that the merge
commit is being updated, which allows us to wait for it before we
return and let the daemon continue.

doudou committed:
fix: disable faraday caching for github

doudou commented:
@g-arjones turned out that there were other issues related to merge commits. I pushed significant changes to fix that, and updated the PR message to reflect what happened. Could you please have a look ?

g-arjones commented on lib/autoproj/daemon/git_api/services/github.rb:
```diff
@@ -53,7 +67,35 @@ def initialize(**options)
+                    # @param [GitAPI::URL] repo_url the repository URL
+                    # @param [Integer] number the pull request number
+                    def pull_request_mergeable?(git_url, number, poll: 0.1)
+                        loop do
```
I think you should add a timeout so that the daemon doesn't hang forever if `mergeable` doesn't become available for whatever reason (ie. github bug). I've personally seen this happen on gitlab. Also, it would be good to add some kind of output so one can keep track of what's happening in the logs (i.e how long it waited for the `mergeable` status if one wasn't available after a reasonable amount of time)

doudou replied:
Ack on the timeout. I should return false then I guess ?

There's the "waiting for ..." message. That's how I found out I had to remove the cache.

g-arjones replied:
> I should return false then I guess ?

Sounds reasonable.

> There's the "waiting for ..." message.

I missed that somehow.

g-arjones commented on lib/autoproj/daemon/git_api/services/github.rb:
```diff
@@ -53,7 +67,35 @@ def initialize(**options)
-                                PullRequest.from_ruby_hash(git_url, pr.to_hash)
+                                mergeable =
+                                    if pr["merged_at"]
+                                        false
```
Why would you use a different setting in this case?

doudou replied:
I ... don't know actually. Will remove.

doudou committed:
feat: add a cache for the mergeability status

Will query the pull_request endpoint only if the base or pr head
have changed.

doudou committed:
doc: document the pull_request argument of Client#extract_info_from_pull_request_ref

Its usefulness is really not obvious from the outside ...

doudou committed:
fix: make git_url a part of the equality / hash implementations of JSONFacade

doudou committed:
fix: do not query the Git services when resolving dependencies

We resolve all pull requests in a single spot, so we can actually
resolve all dependencies there once and for all.

Until now it was not such a big deal since we were using faraday
cache, but now that we disabled it to workaround the merge issues
on github, it has a big impact. Essentially, we are doing one query
per pull request in #pull_requests, plus one per pull request in the
PR dependency tree, including the root. Duplicated if a given PR is
more than once in a dependency tree.

That's a lot of requests.

doudou committed:
fix: do not handle `mergeable` differently for already merged pull requests

doudou committed:
fix(test): return from Timecop in the helper's teardown

Just avoids weird test ordering issues

doudou committed:
fix: add a timeout when trying to determine mergeability

doudou committed:
fix: reenable caching for everything but the direct pull request endpoint

Turns out I was being very optimistic on GH's rate limiting ... It is
in the end borderline for our usage (5000 queries per hour, which is
~80 per minute). Create two clients, one with and one without cache,
which in combination with the mergeability cache should definitely be
enough to not hit any limit at all.

doudou commented:
@g-arjones another round. Turns out this whole thing is a rabbit hole.

GitHub's rate limit is ~80 queries per minute. Our daemon does trigger it - not often, but regularly. The new commits add a mergeability cache, re-enable the HTTP cache for the all endpoints except the pull request one, and refactor dependency handling to not do any client query.

The latter bit is not needed now that I reenabled caching, but since I implemented it I'd like to keep it. I think it makes things a lot clearer (and is a small step towards new functionality I was hoping to implement in the future regarding how dependent builds are handled / represented)

g-arjones commented on lib/autoproj/daemon/git_api/services/github.rb:
```diff
@@ -30,32 +30,145 @@ class GitHub < GitAPI::Service
+                    )
+                        super(**options)
+
+                        @client = create_octokit_client(cache: false)
```
`cache: true` ?

doudou replied:
Yes. Did not run the tests before I pushed :shame:

g-arjones commented on lib/autoproj/daemon/git_api/json_facade.rb:
```diff
@@ -29,15 +29,15 @@ def initialize(git_url, model)
 
                 def ==(other)
-                    @model == other.model
+                    @git_url == other.git_url && @model == other.model
```
Is it possible to have an equal model but different URL?

doudou replied:
In practice, outside of tests, probably not since we use this mostly for pull requests and there are URLs in the model (at least on GitHub).

In theory ? Yes. JSONFacade.new(someurl, {}) != JSONFacade.new(otherurl, {}).

Since it is pretty simple ... I thought that making the tests more robust would not hurt ...

g-arjones commented:
I'm going to have a deeper look into this later but I'm not sure about the new handling of pull request dependencies because:

1. This assumes that all pull requests from all packages are fetched before the dependencies are resolved, which may not be obvious for the reader and may not be the case for features that I also plan for the future (i.e support webhooks);
2. I must be able to generate overrides for packages that are in the workspace but are not being watched by the daemon (a.k.a are not in the list of "known" PRs that you pass to the `OverridesRetriever`)

Plus, with cache enabled I don't think the change has any meaningful impact on performance

doudou commented:
Funny, because I'm going for a completely opposite direction.

I *want* a central repository of all the packages, repositories, pull requests and their dependencies as a graph. The three major features that it would enable are:

- having a single daemon for multiple workspaces without duplicating watches
- bidirectional dependencies (i.e. announce that two pull requests *must* be built together)
- trigger "parent builds" when a dependency changes

To me, this is also a MUCH better design. It's a lot simpler.

And I don't think it's a bad desig with what you want, actually.

1. for the "URLs outside the current workspace": actually, you will have to resolve dependencies early (and not on overrides generation) so that you can inject the "new repos" into the whole process. This PR definitely helps. I'd even say that any design that does not centralize the dependency information will be a nightmarish plate of spaghetti.
2. I don't see the problem there either. This PR's design handles very well that case. Update the dependencies of the pull request object "modified" by the hook and run the rest of the process.

g-arjones commented on test/autoproj/daemon/test_overrides_retriever.rb:
```diff
@@ -116,143 +99,117 @@ def create_pull_request(owner, name, number, body, state: "open")
+                    depends = retriever.dependencies(pr_drivers_gps_ublox)
+                    assert_equal Set[gitlab, github], depends.to_set
                 end
-                it "allows depending on pull requests from different services" do
```
Could you keep this test case, please?

doudou replied:
I kept it. It's just under #dependencies since #retrieve_dependencies does not exist anymore

g-arjones commented on test/autoproj/daemon/test_overrides_retriever.rb:
```diff
@@ -116,143 +99,117 @@ def create_pull_request(owner, name, number, body, state: "open")
                 end
             end
-            describe "#retrieve_dependencies" do
-                it "recursively fetches pull request dependencies" do
```
Could you keep this test case, please?

doudou replied:
Same answer. There is no recursive fetching since the dependency refactoring fetches flat and then PullRequest#recursive_dependencies resolve the recursion.

g-arjones commented on test/autoproj/daemon/test_overrides_retriever.rb:
```diff
@@ -116,143 +99,117 @@ def create_pull_request(owner, name, number, body, state: "open")
+                    assert_equal [pr_driver_orogen_iodrivers_base,
                                   pr_package_set], depends
                 end
-                it "breaks cyclic dependencies" do
```
Could you keep this test case, please?

doudou replied:
Same answer. There is no recursive fetching since the dependency refactoring fetches flat and then PullRequest#recursive_dependencies resolve the recursion.

I'll move these test cases to PullRequest

g-arjones commented on lib/autoproj/daemon/git_api/pull_request.rb:
```diff
@@ -85,6 +96,23 @@ def draft?
+                # Resolve the list of recursive dependencies for this pull request
+                #
+                # @return [Set<GitAPI::PullRequest>]
+                def recursive_dependencies
```
How is this recursive?

doudou replied:
It resolves the 'recursive dependencies'. Which means it resolves the dependencies of the dependencies. Does not mean the algorithm to do so has to be recursive.

doudou commented:
@g-arjones if you still do not want the dependency refactoring, I'll just move it to another PR

g-arjones commented:
> if you still do not want the dependency refactoring, I'll just move it to another PR

No, I had another look and it's fine. It's just that my goal was to move the dependencies resolution to another plugin and generate overrides during build, which would, among other things, allow testing PRs on the buildconf and testing new packages. Having to fetch pull requests from all packages in that context doesn't make sense in my opinion.

But, in the context of the daemon, I think the new design looks good. Sorry for all the noise.

g-arjones commented:
```ruby
                    it "handles cycles in the pull requests dependencies" do
                        pr0 = PullRequest.new("0", {})
                        pr1 = PullRequest.new("1", {})
                        pr2 = PullRequest.new("2", {})

                        pr0.dependencies = [pr1, pr2]
                        pr1.dependencies = [pr2]
                        pr2.dependencies = [pr0]

                        assert_equal Set[pr1, pr2], pr0.recursive_dependencies.to_set
                        assert_equal Set[pr0, pr2], pr1.recursive_dependencies.to_set
                        assert_equal Set[pr0, pr1], pr2.recursive_dependencies.to_set
                    end
```

This should pass, right?

doudou commented:
> This should pass, right?

It does, doesn't it ?

g-arjones commented:
> It does, doesn't it ?

No.

doudou committed:
fix: pass options from the client configuration to the github octokit client

doudou commented:
??? The tests do pass.... Both locally and here. I don't get it.

doudou commented:
> ??? The tests do pass.... Both locally and here. I don't get it.

I mean "I don't get what you're saying". The failure I did have looked like a very weird GH-action specific failure in atomic_write.

g-arjones commented:
> ??? The tests do pass.... Both locally and here. I don't get it.

I mean the exact test case that I sent where the `PullRequest` URLs are different.

```ruby
                    it "handles cycles in the pull requests dependencies" do
                        pr0 = PullRequest.new("0", {})
                        pr1 = PullRequest.new("1", {})
                        pr2 = PullRequest.new("2", {})

                        pr0.dependencies = [pr1, pr2]
                        pr1.dependencies = [pr2]
                        pr2.dependencies = [pr0]

                        assert_equal Set[pr1, pr2], pr0.recursive_dependencies.to_set
                        assert_equal Set[pr0, pr2], pr1.recursive_dependencies.to_set
                        assert_equal Set[pr0, pr1], pr2.recursive_dependencies.to_set
                    end
```

doudou committed:
fix: test and implementation of PullRequest#recursive_dependencies

The test was using PullRequest objects that were actually equivalent
from the p.o.v. of Hash, which led to it passing even though it was
broken.

Fix the test by passing different URLs to each PullRequest object, and
then the implementation

doudou commented:
Ah .... right ... Did not notice it was different from the actual test case.

Anyways, fixed. Thanks for catching this !

g-arjones commented:
Now I see recursion. I was really puzzled how this thing would work if you were never calling `#dependencies` on the dependencies... :+1: 

Looks good to me.

g-arjones commented on lib/autoproj/daemon/git_api/pull_request.rb:
```diff
@@ -85,6 +96,23 @@ def draft?
+                        next if result.include?(pr) || pr == self
+
+                        result << pr
+                        queue.concat(pr.dependencies)
```
Actually, shouldn't it be 

```ruby
                        queue.concat(pr.recursive_dependencies)
```

?

g-arjones replied:
Ah, doesn't matter, you add them to the queue. Nevermind.

doudou replied:
That would be doing actual programming recursion, and we would have to pass the "seen" set to the method to avoid infinite loops. In these cases (graphs with loops), I usually prefer doing a breadth-first search (which I'm doing here, that's what the `queue` is here for). I find that clearer personally.

g-arjones replied:
Yep, agreed.
